documentation/ErrorStop.md

```diff
@@ -0,0 +1,103 @@
+
+2. Canonicalize DO statements
+
+3. Resolve names (produce *tree of scopes* populated with symbols and types)
+
+4. Rewrite Parse Tree
+
+5. Statement semantics analysis pass 1 (see below)
+
+6. Statement semantics analysis pass 2 (see below)
+
+7. Run Mod File Writer
+
+Pass 1 of the Statements Semantic Analysis currently analyzes experssions (semantic checks performed on all expressions):
+
+~~~
+using StatementSemanticsPass1 = SemanticsVisitor<ExprChecker>;
+~~~
+
+Pass 2 of the Statements Semantic Analysis currently analyzes assignments (semantic checks performed on all assignment statements) and other statements and language constructs (e.g. DO CONCURRENT); all those checkers are listed in alphabetic order:
+
+~~~
+using StatementSemanticsPass2 =
+    SemanticsVisitor</* list of Pass2 checkers ...*/>;
+~~~
+
+Each of those checkers is defined as a class inheriting from `BaseChecker`. In order to implement their activities, they can override `Enter()` and/or `Leave()` methods that are no-op (i.e. `{}`) in the `BaseChecker` class. These methods are called for the parse tree nodes of interest, `Ender()` before the children, `Leave()` after.
+
+* `ExprChecker` overrides `Entry()` method for `Expr` and `Variable` parser nodes.
+* `AssignmentChecker` overrides `Entry()` method for `AssignmentStmt`, `PointerAssignmentStmt`, `WhereStmt`, `WhereConstruct`, `ForallStmt`, `ForallConstruct` parser nodes.
+* `DoConcurrentChecker` overrides `Leave()` method for `DoConstruct` parser nodes.
+
+Both of these passes are instances of `SemanticVisitor` on which `Walk()` method is applied, which in turn calls `parser::Walk()` method:
+
+~~~
+template<typename... C> class SemanticsVisitor : public virtual C... {
+public:
+  using C::Enter...;
+  using C::Leave...;
+  using BaseChecker::Enter;
+  using BaseChecker::Leave;
+  SemanticsVisitor(SemanticsContext &context)
+    : C{context}..., context_{context} {}
+  template<typename N> bool Pre(const N &node) {
+    Enter(node);
+    return true;
+  }
+  template<typename N> void Post(const N &node) { Leave(node); }
+  bool Walk(const parser::Program &program) {
+    parser::Walk(program, *this);
+    return !context_.AnyFatalError();
+  }
+
+private:
+  SemanticsContext &context_;
+};
+~~~
+
+As we can see, whenever the visitor's `Pre()` method is called for any given node of the parse tree, the `Enter()` method of a given checker is executed for that node. Similarly, whenever the visitor's `Post()` method is called for any given node of the parse tree, the `Leave()` method of a given checker is executed for that node.
+
+In order to prevent code duplications, useful functions (to be called from within the checkers) were extracted into `tools.cc` file and exposed to te `semantics` namespace. E.g. `ExprChecker` makes good use of `semantics::FindExternallyVisibleObject()` function.
+
+## Proposed checker of ERROR STOP statements
```

Comment:
I think everything before line 91 is fairly generic.  I think it can be deleted.  If you think some of it should be preserved, perhaps you could move it to the documentation for semantics.